storage/scylla_api.c

```diff
@@ -0,0 +1,707 @@
+  }
+  for (size_t i = 0; i < trans_num; i++) {
+    statement = ret_insert_bundle_statement(insert_prepared, transaction + i);
+    if (execute_statement(session, statement) != CASS_OK) {
+      ret = SC_STORAGE_CASSANDRA_QUREY_FAIL;
+      goto exit;
+    }
+  }
+
+exit:
+  cass_prepared_free(insert_prepared);
+  return ret;
+}
+
+status_t insert_transaction_into_edgeTable(CassSession* session, scylla_iota_transaction_t* transaction,
+                                           size_t trans_num) {
+  status_t ret = SC_OK;
+  const CassPrepared* insert_prepared = NULL;
+  CassStatement* statement = NULL;
+  const char* insert_query =
+      "INSERT INTO permanent.edgeTable (edge, bundle, address, hash)"
+      "VALUES (?, ?, ?, ?);";
+
+  if (transaction == NULL || trans_num == 0) {
+    return SC_STORAGE_INVAILD_INPUT;
+  }
+
+  if (prepare_query(session, insert_query, &insert_prepared) != CASS_OK) {
+    ta_log_error("%s\n", "prepare INSERT query fail");
+    return SC_STORAGE_CASSANDRA_QUREY_FAIL;
+  }
+
+  for (size_t i = 0; i < trans_num; i++) {
+    statement = ret_insert_edge_statement(insert_prepared, transaction + i, (transaction + i)->address);
+    if (execute_statement(session, statement) != CASS_OK) {
+      ret = SC_STORAGE_CASSANDRA_QUREY_FAIL;
+      goto exit;
+    }
+    statement = ret_insert_edge_statement(insert_prepared, transaction + i, (transaction + i)->trunk);
+    if (execute_statement(session, statement) != CASS_OK) {
+      ret = SC_STORAGE_CASSANDRA_QUREY_FAIL;
+      goto exit;
+    }
+    statement = ret_insert_edge_statement(insert_prepared, transaction + i, (transaction + i)->branch);
+    if (execute_statement(session, statement) != CASS_OK) {
+      ret = SC_STORAGE_CASSANDRA_QUREY_FAIL;
+      goto exit;
+    }
+  }
+
+exit:
+  cass_prepared_free(insert_prepared);
+  return ret;
+}
+
+status_t select_from_bundleTable(CassSession* session, CassStatement* statement,
+                                 scylla_iota_transaction_t** transaction, size_t* transactionNum) {
+  int idx;
+  status_t ret = SC_OK;
+  CassFuture* future = NULL;
+  const CassResult* result;
+  CassIterator* iterator;
+
+  future = cass_session_execute(session, statement);
+  cass_future_wait(future);
+  if (cass_future_error_code(future) != CASS_OK) {
+    print_error(future);
+    ret = SC_STORAGE_CASSANDRA_QUREY_FAIL;
+    goto exit;
+  }
+
+  /* get transaction number */
+  result = cass_future_get_result(future);
+  iterator = cass_iterator_from_result(result);
+  *transactionNum = 0;
+  while (cass_iterator_next(iterator)) {
+    (*transactionNum)++;
+  }
+  *transaction = malloc((*transactionNum) * sizeof(scylla_iota_transaction_t));
+  if (*transaction == NULL) {
+    ta_log_error("%s\n", "malloc failure");
+    ret = SC_STORAGE_OOM;
+    goto end_iterate;
+  }
+  /* get each transaction result */
+  idx = 0;
+  iterator = cass_iterator_from_result(result);
+  while (cass_iterator_next(iterator)) {
```

Comment:
while not use for loor ?

Reply:
It will be as following
`for(cass_bool_t is_next_itr = cass_iterator_next(iterator) , int idx = 0; is_next_itr ;  is_next_itr = cass_iterator_next(iterator) , idx++)
  `

Reply:
I personally don't like the declaration of `idx` which is far away from the place it is used, so I like the implementation you provide here more. However, it is kinda long ><